mlflow/pyfunc/__init__.py

```diff
@@ -1511,6 +1516,25 @@ def udf(
         # we can load the model from correct path
         mlflow.set_tracking_uri(tracking_uri)
 
+        @keras_custom_object_scope(custom_objects)
+        def load_model_for_local_env_manager():
+            if is_spark_connect and not should_spark_connect_use_nfs:
+                model_path = os.path.join(
+                    tempfile.gettempdir(),
+                    "mlflow",
+                    insecure_hash.sha1(model_uri.encode()).hexdigest(),
+                )
+                try:
+                    loaded_model = mlflow.pyfunc.load_model(model_path)
+                except Exception:
+                    os.makedirs(model_path, exist_ok=True)
+                    loaded_model = mlflow.pyfunc.load_model(model_uri, dst_path=model_path)
+            elif should_use_spark_to_broadcast_file:
+                loaded_model, _ = SparkModelCache.get_or_load(archive_path)
+            else:
+                loaded_model = mlflow.pyfunc.load_model(local_model_path)
+            return loaded_model
```

Comment:
If we want to decorate model loading with `@keras_custom_object_scope(custom_objects)`,

why not doing it inside `mlflow.sklearn.load_model` ?
so that we can avoid changing code in `spark_udf` and making the code easier to maintain.

Reply:
Because the custom_objects is on driver, we need to pass it to workers. We don't know this parameter if we put the logic within load_model.

Reply:
Got it, make sense!

Reply:
Issue:

For the case env_manger != "local", it uses "scoring server" to load the model, but you does not set up custom objects in this case. Does the unit tests cover the case ?

Refactoring suggestion:

We can define 2 functions `get_model_custom_objects` `with_model_custom_objects` in `mlflow.tensorflow` module, and in spark_udf, for specific flavor, we check whether `get_model_custom_objects` `with_model_custom_objects` function exists, if yes then we call it.

so that we can decouple keras specific code with spark_udf code.

and we should also make "scoring server" supporting this.

Reply:
Yes we need to fix scoring server as well.
I don't understand the refactoring suggestion, now we have `get_custom_objects` in tensorflow module, and `keras_custom_object_scope` in annotations, because custom_object_scope can only be used as a context manager, we need to wrap the model loading function within it. What's the difference than the current change? 🤔

Reply:
Updated: I save custom_objects into file so that we don't need to worry about sending them in spark UDF or serving.

Reply:
Saving `custom_objects` is good, we don't need above APIs :)